### PR TITLE
dsa: Add e2e tests

### PR DESCRIPTION
**Pull request by ozhuraki**
Closes: #1026 
Closes: #642

### CODE REVIEW

File: test/e2e/dsa/dsa.go [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+		framework.RunKubectlOrDie(f.Namespace.Name, "--namespace", f.Namespace.Name, "apply", "-k", filepath.Dir(kustomizationPath))
+
+		ginkgo.By("waiting for DSA plugin's availability")
+		if _, err := e2epod.WaitForPodsWithLabelRunningReady(f.ClientSet, f.Namespace.Name,
+			labels.Set{"app": "intel-dsa-plugin"}.AsSelector(), 1 /* one replica */, 100*time.Second); err != nil {
+			framework.DumpAllNamespaceInfo(f.ClientSet, f.Namespace.Name)
+			kubectl.LogFailedContainers(f.ClientSet, f.Namespace.Name, framework.Logf)
+			framework.Failf("unable to wait for all pods to be running and ready: %v", err)
+		}
```
Comment (bart0sh) [inline test/e2e/dsa/dsa.go:63]: Here we should at least check if DSA resource(s) are available on the node.

**Review by bart0sh — CHANGES_REQUESTED** [review]
Check if plugin announces DSA resources properly

**Comment by mythi** [comment]
> Check if plugin announces DSA resources properly

Also, prepare with changes required from #985 (i.e., pod security admission)

**Comment by bart0sh** [comment]
> > Check if plugin announces DSA resources properly
> 
> Also, prepare with changes required from #985 (i.e., pod security admission)

And add running[ test workload](https://github.com/intel/intel-device-plugins-for-kubernetes/tree/main/demo/dsa-accel-config-demo).

**Comment by ozhuraki** [comment]
@bart0sh 
> Here we should at least check if DSA resource(s) are available on the node.

This is checked now, the demo scheduling depends on the DSA resources.

@mythi 
> Also, [...] (i.e., pod security admission

Done

@bart0sh @mythi 

Updated, take a look please

File: test/e2e/dsa/dsa.go (hunk 1 above)
Comment (mythi) [reply test/e2e/dsa/dsa.go:81]: This is still missing. For others we have, e.g.,
https://github.com/intel/intel-device-plugins-for-kubernetes/blob/1e6da3001f6989c2b14e97286f7668aae604a51f/test/e2e/qat/qatplugin_dpdk.go#L77-L80

File: demo/dsa-accel-config-demo-pod.yaml [hunk 2]
```diff
@@ -2,15 +2,36 @@ apiVersion: v1
...
+  labels:
+    app: dsa-accel-config-demo
 spec:
   containers:
   - name: dsa-accel-config-demo
-    image: dsa-accel-config-demo:devel
+    image: intel/dsa-accel-config-demo:devel
+    command: [ "/bin/bash", "-e", "-c", "cd /test; ./dsa_test -v -w 0 -d dsa0/wq0.0 -f 0x0 -o 0x3; while true; do sleep 300000; done" ]
+    securityContext:
+      privileged: true
```
Comment (mythi) [inline demo/dsa-accel-config-demo-pod.yaml:13]: some debug leftover?

File: demo/dsa-accel-config-demo/idxd-reset.patch [hunk 3]
```diff
@@ -0,0 +1,26 @@
+From 07e1137a71021dc02df98b942a260b70ff852ca5 Mon Sep 17 00:00:00 2001
+From: Oleg Zhurakivskyy <oleg.zhurakivskyy@intel.com>
+Date: Tue, 3 May 2022 15:28:05 +0300
+Subject: [PATCH] idxd: Ignore reset device error
```
Comment (mythi) [inline demo/dsa-accel-config-demo/idxd-reset.patch:4]: it's not clear why this is needed. have we reported this bug to `idxd-config` upstream?

File: deployments/dsa_plugin/overlays/dsa_initcontainer/dsa_initcontainer.yaml [hunk 4]
```diff
@@ -17,13 +17,18 @@ spec:
           readOnlyRootFilesystem: true
           privileged: true
         volumeMounts:
+        - mountPath: /sys/bus/dsa
+          name: sys-bus-dsa
```
Comment (mythi) [inline deployments/dsa_plugin/overlays/dsa_initcontainer/dsa_initcontainer.yaml:21]: do we need it for the operator too?

**Review by mythi — COMMENTED** [review]
it looks this PR needs to be split into three separate PRs: 1) YAML fixes, 2) `test/e2e/dsa` (this PR), and 3) github workflows part

File: demo/dsa-accel-config-demo-pod.yaml (hunk 2 above)
Comment (ozhuraki) [reply demo/dsa-accel-config-demo-pod.yaml]: Actually no, seems like this is required at least for the bmetal. It's unclear to me if this ever worked (bisecting, cross-checking)

File: demo/dsa-accel-config-demo/Dockerfile [hunk 5]
```diff
@@ -1,9 +1,10 @@
-FROM fedora:32 AS builder
+FROM fedora:35 AS builder
 
-RUN dnf install -y wget autoconf automake libtool m4 diffutils file make dnf-utils
-RUN dnf install -y libuuid-devel json-c-devel kmod-devel libudev-devel
+RUN dnf install -y \
+    wget autoconf automake libtool m4 diffutils file make dnf-utils patch \
+    libuuid-devel json-c-devel kmod-devel libudev-devel
```
Comment (mythi) [inline demo/dsa-accel-config-demo/Dockerfile:5]: was it so that `kmod` and `udev` where not needed?
```suggestion
    libuuid-devel json-c-devel
```

Comment (ozhuraki) [reply demo/dsa-accel-config-demo/Dockerfile]: Yes, both not needed, dropped them

**Review by mythi — DISMISSED** [review]

**Review by bart0sh — DISMISSED** [review]
/lgtm

@mythi please, don't merge yet. We need to test it first.

**Comment by ozhuraki** [comment]
@bart0sh @mythi 
Changes as discussed, please take a look:
- Added idxd reset patch to the dsa demo
- Added an overlay for the dsa demo for the bmetal environment

**Review by mythi — APPROVED** [review]

**Comment by mythi** [comment]
note: `dsa-accel-config-demo` is not yet built by CI because the job gets cancelled early.

File: demo/dsa-accel-config-demo/idxd-reset.patch [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+diff --git a/accfg/lib/libaccfg.c b/accfg/lib/libaccfg.c
+index be3be69..32eb71f 100644
+--- a/accfg/lib/libaccfg.c
++++ b/accfg/lib/libaccfg.c
+@@ -636,7 +636,7 @@ static void *add_device(void *parent, int id, const char *ctl_base,
+ 
+ 	rc = accfg_set_param(ctx, dfd, "cmd_status", "1", 1);
+ 	/* older drivers don't support writing to cmd_status */
+-	if (rc && rc != -EACCES) {
++	if (0 && rc && rc != -EACCES) {
```
Comment (bart0sh) [inline demo/dsa-accel-config-demo/idxd-reset.patch:20]: Does this mean we ignore accfg_set_param return code?

File: test/e2e/dsa/dsa.go [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+	if err != nil {
+		framework.Failf("unable to locate %q: %v", configmapYaml, err)
+	}
+
+	demoPath, err := utils.LocateRepoFile(demoYaml)
+	if err != nil {
+		framework.Failf("unable to locate %q: %v", demoYaml, err)
+	}
+
+	ginkgo.It("checks availability of DSA resources", func() {
```
Comment (bart0sh) [inline test/e2e/dsa/dsa.go:59]:
```suggestion
        ginkgo.It("runs DSA plugin and a demo workload", func() {
```

File: test/e2e/dsa/dsa.go [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+
+		ginkgo.By("checking the resource is allocatable")
+		if err := utils.WaitForNodesWithResource(f.ClientSet, "dsa.intel.com/wq-user-dedicated", 300*time.Second); err != nil {
+			framework.Failf("unable to wait for nodes to have positive allocatable resource: %v", err)
+		}
+
+		framework.RunKubectlOrDie(f.Namespace.Name, "--namespace", f.Namespace.Name, "apply", "-k", filepath.Dir(demoPath))
+
+		ginkgo.By("waiting for the DSA demo to succeed")
+		f.PodClient().WaitForSuccess("dsa-accel-config-demo", 300*time.Second)
```
Comment (bart0sh) [inline test/e2e/dsa/dsa.go:81]: Can we add something like this here? https://github.com/intel/intel-device-plugins-for-kubernetes/blob/main/test/e2e/gpu/gpu.go#L97 to see the test output or its tail if it's too long?

**Review by mythi — APPROVED** [review]
LGTM, thanks!

**Comment by bart0sh** [comment]
/lgtm

@ozhuraki can we add .github/workflows/e2e-dsa.yml to this PR to see if dsa test passes?

**Comment by ozhuraki** [comment]
@mythi @bart0sh 
Changes, please take a look:
- Dropped privileged for the demo
- Dropped customization overlay for bmetal CI
- Added a log output for the demo
- Refined idxd-reset.patch (needed for kernels 5.17.x)
- Extended the demo coverage

**Comment by mythi** [comment]
> @ozhuraki can we add .github/workflows/e2e-dsa.yml to this PR to see if dsa test passes?

I asked to keep that separate so that the two commits are split.

**Comment by bart0sh** [comment]
@mythi how can we see if the DSA CI job working then?

**Comment by mythi** [comment]
> @mythi how can we see if the DSA CI job working then?

let's then cherry-pick the commit from #1026 and have two commits here. Normally the e2e test case development (e.g. this PR) has been independent from setting up the CI (#1026) but this time it's a rather special case.

**Comment by ozhuraki** [comment]
Changes:
- Added a CI workflow for e2e DSA

**Review by mythi — APPROVED** [review]

**Pull request merged**